PetGame/Services/ActivityService.cs

```diff
@@ -0,0 +1,170 @@
+                        };
+                        ret.Add(n);
+                    }
+                }
+                // return all of the results
+                return ret;
+            }
+        }
+
+        /// <summary>
+        ///     Gets an activity by Id.
+        /// </summary>
+        /// <param name="activityId">the ActivityId to get</param>
+        /// <returns>
+        ///     The value of the Activity, or null if not found.
+        /// </returns>
+        public Activity GetActivityById(ulong activityId)
+        {
+            Activity ret = null;
+            using (var conn = sqlManager.EstablishDataConnection)
+            {
+                var cmd = conn.CreateCommand();
+                cmd.CommandText =
+                    @"SELECT ActivityId, PetId, Timestamp, Type FROM Activity
+                        FROM Activity
+                        WHERE ActivityId = @ActivityId;";
+                cmd.Parameters.AddWithValue("@ActivityId", $"{activityId}");
+
+                // read the results
+                using (var reader = cmd.ExecuteReader())
+                {
+                    while (reader.Read())
+                    {
+                        ret = new Activity()
+                        {
+                            ActivityId = (ulong)reader.GetInt64(0),
+                            PetId = (ulong)reader.GetInt64(1),
+                            Timestamp = reader.GetDateTime(2),
+                            Type = (ActivityType)reader.GetChar(3)
+                        };
+                    }
+                }
+                return ret;
+            }
+        }
+
+        /// <summary>
+        ///     Inserts a new Activity into the database.
+        /// </summary>
+        /// <param name="activity"> The activity to insert. </param>
+        /// <returns>
+        ///     The activity object supplied with the ActivityId property set.
+        /// </returns>
+        public Activity InsertActivity(Activity activity)
+        {
+            if (activity == null)
+                throw new ArgumentNullException(paramName: nameof(activity),
+                    message: "The supplied Activity cannot be null.");
+
+            using (var conn = sqlManager.EstablishDataConnection)
+            {
+                var cmd = conn.CreateCommand();
+                cmd.CommandText =
+                    @"INSERT INTO Activity (PetId, Timestamp, Type)
+                      OUTPUT INSERTED.ActivityId 
+                      VALUES (@PetId, @Timestamp, @Type);";
+                cmd.Parameters.AddWithValue("@PetId", $"{activity.PetId}");
+                cmd.Parameters.AddWithValue("@Timestamp", activity.Timestamp);
+                cmd.Parameters.AddWithValue("@Type", (char)activity.Type);
+
+                // read the results
+                using (var reader = cmd.ExecuteReader())
+                {
+                    while (reader.Read())
```

Comment:
Unhandled Exception occurs if the Pet does not exist in the DB. Expected Behavior?

Reply:
This is expected, and for now intended, because PetId has a foreign key constraint. This same behavior happens for other endpoints too, and should be fixed before the final submission, but isn't a priority right now.